File: K_WeakestRow/weakest.js
/**
 * @param {number[][]} mat
 * @param {number} k
 * @returns {number[]}
 */
var kWeakestRows = function (mat, k) {
	const rowInfo = [];

	for (let i = 0; i < mat.length; i++) {
		rowInfo.push({ rowIndex: i, soldiers: countSoldiers(mat[i]) });
	}

	rowInfo.sort((a, b) => {
		if (a.soldiers === b.soldiers) {
			return a.rowIndex - b.rowIndex;
		}
		return a.soldiers - b.soldiers;
	});

	const result = [];

	for (let i = 0; i < k; i++) {
		result.push(rowInfo[i].rowIndex);
	}

	return result;
};

/**
 * @param {number[]} row
 * @returns {number}
 */
const countSoldiers = (row) => {
	let [i, j] = [0, row.length - 1];

	while (i <= j) {
		let mid = Math.floor((i + j) / 2);
		if (row[mid] == 1) {
			i = mid + 1;
		} else {
			j = mid - 1;
		}
	}

	return i;
};

File: K_WeakestRow/weakest.py
from typing import List


class Solution:
    def kWeakestRows(self, mat: List[List[int]], k: int) -> List[int]:
        row_info: List[RowInfo] = []

        i = 0
        while i < len(mat):
            row_info.append(RowInfo(i, count_soldiers(mat[i])))
            i += 1

        row_info = sorted(row_info, key=lambda x: (x.soldiers, x.row_index))

        weakest_rows = [info.row_index for info in row_info[:k]]
        return weakest_rows


class RowInfo:
    def __init__(self, row_index: int, soldiers: int):
        self.row_index = row_index
        self.soldiers = soldiers


def count_soldiers(row: List[int]) -> int:
    i, j = 0, len(row) - 1

    while i <= j:
        mid = (i + j) // 2

        if row[mid] == 1:
            i = mid + 1
        else:
            j = mid - 1

    return i
